packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts

```diff
@@ -0,0 +1,138 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as browserslist from 'browserslist';
+import * as path from 'path';
+import { ScriptTarget } from 'typescript';
+
+const es6ModuleSupport = require('./es6-module-support.json');
+const supportTable = convertSupportTableToBrowserVersionTuples(es6ModuleSupport);
+
+export function isDifferentialLoadingNeeded(
+  tsConfigFileName: string, query: string[] | undefined = undefined) {
+  const es5SupportNeeded = isEs5BrowserSupportNeeded(tsConfigFileName, query);
+  const target = getTargetVersion(tsConfigFileName);
+
+  return es5SupportNeeded && target !== ScriptTarget.ES5;
+}
+
+export function getTargetVersion(tsConfigFileName: string): ScriptTarget {
+  return _getTargetVersion(tsConfigFileName, new Set<string>());
+}
+
+function _getTargetVersion(tsConfigFileName: string, visited: Set<string>): ScriptTarget {
+  if (visited.has(tsConfigFileName)) {
+    throw new Error('cycle in tsconfig.json inheritance detected: '
+      + Array.from(visited).join(', ') );
+  }
+  visited.add(tsConfigFileName);
+
+  const content = require(tsConfigFileName);
+  const compilerOptions = content['compilerOptions'];
+  const target = compilerOptions['target'] as string;
+  const extendsField = content['extends'];
+
+  if (compilerOptions && target) {
+
+    const targetUC = target.toUpperCase();
+
+    // @ts-ignore
+    const scriptTarget = ScriptTarget[targetUC] as ScriptTarget;
+
+    if (typeof scriptTarget === 'undefined') {
+      throw new Error('Unsupported script target in tsconfig.json found: ' + target);
+    }
+
+    return scriptTarget;
+  } else if (extendsField) {
+    const tsConfigDirName = path.dirname(tsConfigFileName);
+    const nextTsConfigFileName = path.join(tsConfigDirName, extendsField);
+
+    return _getTargetVersion(nextTsConfigFileName, visited);
+  } else {
+    // Fallback to ES5
+    return ScriptTarget.ES5;
+  }
+
+}
+
+export function isEs5BrowserSupportNeeded(
```

Comment:
Some docs would help here. Example what is a `query`?

Reply:
+1

Reply:
I'll rename the variable to `browserslistQuery`.